### DISCUSSION TITLE
Switch to config objects for operators with a lot of arguments.

### DISCUSSION THREAD

**Discussion by benlesh** [ReactiveX/rxjs#6331] [General]
There are a few operators that have a lot of arguments, one example would be `groupBy`. This hurts readability a bit and also bites us later if we want to change or deprecate arguments.

So the idea is simple, introduce non-breaking changes to operators like `groupBy` to allow the passing of a config object to essentially provide "named arguments":

```ts
groupBy(fn1, fn2 fn3, fn4)

// becomes

groupBy({
  keySelector: fn1,
  resultSelector: fn2,
  durationSelector: fn3,
  subjectFactory: fn4,
})
```
(The names can be bikeshedded later, ideally to make it more terse but still readable)

**Comment by BioPhoton** [comment]
Hi @benlesh!

If this is still a thing I would create a list of all the operators that have enough arguments to get refactored to a config object.

I would collect: 
- their name
- link to the source file
- their argument names and roughly the types
- suggestion for future naming (just as a placeholder. The real name is up to people that are responsible for it)
- structure of the future config object

I would provide it as markdown like this:

- [ ] [groupBy](https://github.com/ReactiveX/rxjs/blob/master/src/internal/operators/groupBy.ts)
  **Current Arguments:**
  - keySelector: Function
  - elementSelector: Function
  - durationSelector: Function
  - subjectSelector: Function

  **Future Config Object**
  - Name: `groupingConfig`
  - Structure: 
  ```typescript
  {
    keySelector: Function
    elementSelector: Function
    durationSelector: Function
    subjectSelector: Function
  }
  ````
- [ ] [otherOperator](link) 
  **Current Arguments:**
  - ...   

  **Future Config Object**
  - ...

Let me know if the suggested approach would be helpful or if I can adapt anything.

**Comment by benlesh** [comment]
Thanks @BioPhoton.... we probably don't even need that level of detail. Just a list of the operators that have 3 or more arguments would be a great start.

**Comment by BioPhoton** [comment]
Could you suggest a naming pattern?
I.e. having `Config` as postfix or similar?

**Comment by BioPhoton** [comment]
Hi @benlesh!

Here the list of operators/functions that have 3 or more arguments (n means infinite)

**Subjects**
- [ ] ReplaySubject (3)

**Creation Methode**
- [ ] bindCallback (3)
- [ ] bindNodeCallback (3)
- [ ] combineLatest (n)
- [ ] concat (n)
- [ ] forkJoin (n)
- [ ] fromEvent (4)
- [ ] fromEventPattern (4)
- [ ] iif (3)
- [ ] merge (n)
- [ ] of (n)
- [ ] onErrorResumeNext (n)
- [ ] race (n)
- [ ] range (3)
- [ ] timer (3)
- [ ] zip (n)

**Operators**
- [ ] combineLatest (n)
- [ ] concat (n)
- [ ] count (3)
- [ ] endWith (n)
- [ ] every (3)
- [ ] expand (3)
- [ ] find (3)
- [ ] findIndex (3)
- [ ] groupBy (4)
- [ ] merge (n)
- [ ] mergeMap (3)
- [ ] mergeMapTo (3)
- [ ] mergeScan (3)
- [ ] onErrorResumeNext (n)
- [ ] pluck (n)
- [ ] publishReplay (3)
- [ ] race (n)
- [ ] shareReplay (3)
- [ ] single (3)
- [ ] startWith (n)
- [ ] tap (3)
- [ ] throttleTime (3)
- [ ] timeoutWith (3)
- [ ] windowTime (4)
- [ ] withLatestFrom (n)
- [ ] zip (n)
